src/Nethermind/Nethermind.Merge.AuRa/Shutter/ShutterP2P.cs

```diff
@@ -0,0 +1,218 @@
+using Nethermind.Libp2p.Protocols;
+using Nethermind.Blockchain;
+using System;
+using System.Threading.Tasks;
+using System.Threading;
+using System.Collections.Concurrent;
+using Microsoft.Extensions.DependencyInjection;
+using System.Linq;
+using System.Collections.Generic;
+using Nethermind.Crypto;
+using Multiformats.Address;
+using Nethermind.Core;
+using Google.Protobuf;
+using Nethermind.Abi;
+using Nethermind.Consensus.AuRa.Config;
+using Nethermind.Logging;
+using Nethermind.Consensus.Processing;
+using ILogger = Nethermind.Logging.ILogger;
+
+namespace Nethermind.Merge.AuRa.Shutter;
+
+public class ShutterP2P
+{
+    private readonly Func<Dto.DecryptionKeys, bool> _shouldProcessDecryptionKeys;
+    private readonly Action<Dto.DecryptionKeys> _onDecryptionKeysValidated;
+    private readonly IReadOnlyBlockTree _readOnlyBlockTree;
+    private readonly IAbiEncoder _abiEncoder;
+    private readonly ILogger _logger;
+    private readonly ulong InstanceID;
+    private ShutterEonInfo _eonInfo;
+
+    public ShutterP2P(Action<Dto.DecryptionKeys> onDecryptionKeysValidated, Func<Dto.DecryptionKeys, bool> shouldProcessDecryptionKeys, IReadOnlyBlockTree readOnlyBlockTree, ReadOnlyTxProcessingEnvFactory readOnlyTxProcessingEnvFactory, IAbiEncoder abiEncoder, IAuraConfig auraConfig, ILogManager logManager)
+    {
+        _onDecryptionKeysValidated = onDecryptionKeysValidated;
+        _shouldProcessDecryptionKeys = shouldProcessDecryptionKeys;
+        _readOnlyBlockTree = readOnlyBlockTree;
+        _abiEncoder = abiEncoder;
+        _logger = logManager.GetClassLogger();
+        _eonInfo = new(readOnlyBlockTree, readOnlyTxProcessingEnvFactory, abiEncoder, auraConfig, _logger);
+        InstanceID = auraConfig.ShutterInstanceID;
+
+        ServiceProvider serviceProvider = new ServiceCollection()
+            .AddLibp2p(builder => builder)
+            .AddSingleton(new IdentifyProtocolSettings
+            {
+                ProtocolVersion = auraConfig.ShutterP2PProtocolVersion,
+                AgentVersion = auraConfig.ShutterP2PAgentVersion
+            })
+            // .AddLogging(builder =>
+            //     builder.SetMinimumLevel(Microsoft.Extensions.Logging.LogLevel.Trace)
+            //     .AddSimpleConsole(l =>
+            //     {
+            //         l.SingleLine = true;
+            //         l.TimestampFormat = "[HH:mm:ss.FFF]";
+            //     })
+            // )
+            .BuildServiceProvider();
+
+        IPeerFactory peerFactory = serviceProvider.GetService<IPeerFactory>()!;
+        ILocalPeer peer = peerFactory.Create(new Identity(), "/ip4/0.0.0.0/tcp/" + auraConfig.ShutterP2PPort);
+        if (_logger.IsInfo) _logger.Info($"Started Shutter P2P: {peer.Address}");
+        PubsubRouter router = serviceProvider.GetService<PubsubRouter>()!;
+
+        ITopic topic = router.Subscribe("decryptionKeys");
+        ConcurrentQueue<byte[]> msgQueue = new();
+
+        topic.OnMessage += (byte[] msg) =>
```

Comment:
You also need to unsubscribe when the object reaches the end of it's lifecycle. Otherwise it can result in unexpected behavior.